dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs

```diff
@@ -54,55 +58,108 @@ public HandlebarsPlanner(IKernel kernel, HandlebarsPlannerConfig? config = defau
+        planTemplate = planTemplate.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
+        planTemplate = planTemplate.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
 
-        template = MinifyHandlebarsTemplate(template);
-        return new HandlebarsPlan(this._kernel, template);
+        planTemplate = MinifyHandlebarsTemplate(planTemplate);
+        return new HandlebarsPlan(this._kernel, planTemplate, createPlanPrompt);
     }
 
     private List<FunctionView> GetAvailableFunctionsManual(CancellationToken cancellationToken = default)
     {
-        return this._kernel.Functions.GetFunctionViews()
+        var availableFunctions = this._kernel.Functions.GetFunctionViews()
             .Where(s => !this._config.ExcludedPlugins.Contains(s.PluginName, StringComparer.OrdinalIgnoreCase)
                 && !this._config.ExcludedFunctions.Contains(s.Name, StringComparer.OrdinalIgnoreCase)
                 && !s.Name.Contains("Planner_Excluded"))
-            .ToList();
+                .ToList();
+
+        var functionsView = new List<FunctionView>();
+        foreach (var skFunction in availableFunctions)
+        {
+            // Extract any complex schemas for isolated render in prompt template
+            var parametersView = new List<ParameterView>();
+            foreach (var parameter in skFunction.Parameters)
+            {
+                var paramToAdd = this.SetComplexTypeDefinition(parameter);
+                parametersView.Add(paramToAdd);
+            }
+
+            var returnParameter = skFunction.ReturnParameter.ToParameterView(skFunction.Name);
+            returnParameter = this.SetComplexTypeDefinition(returnParameter);
+
+            // Need to override function view in case parameter views changed (e.g., converted primitive types from schema objects)
+            var functionView = new FunctionView(skFunction.Name, skFunction.PluginName, skFunction.Description, parametersView, returnParameter.ToReturnParameterView());
+            functionsView.Add(functionView);
+        }
+
+        return functionsView;
+    }
+
+    // Extract any complex schemas for isolated render in prompt template
+    private ParameterView SetComplexTypeDefinition(ParameterView parameter)
+    {
+        // TODO (@teresaqhoang): Handle case when schema and ParameterType can exist i.e., when ParameterType = RestApiResponse
+        if (parameter.ParameterType is not null)
+        {
+            // Async return type - need to extract the actual return type and override ParameterType property
+            var type = parameter.ParameterType;
+            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>))
+            {
+                parameter = parameter with { ParameterType = type.GenericTypeArguments[0] }; // Actual Return Type
+            }
+
+            this._parametersTypeView.UnionWith(parameter.ParameterType.ToHandlebarsParameterTypeView());
```

Comment:
With this class member, it looks like the planner object of this class is expected to be discarded after a plan is created. Is this correct?

Reply:
Yep! It's only needed for prompt rendering

Reply:
Synced offline -- removed complex type definitions as class properties. These views now only exist in request scope of CreatePlanAsync